### ISSUE TITLE
Date deserialization fails if type deserializer is present

### ISSUE THREAD

**Issue by aseovic** [FasterXML/jackson-databind#1103]
I just ran into a corner case bug with `java.util.Date` deserialization:

```
com.fasterxml.jackson.databind.JsonMappingException: Unexpected token (VALUE_STRING), expected START_ARRAY: need JSON Array to contain As.WRAPPER_ARRAY type information for class java.util.Date
 at [Source: java.io.BufferedInputStream@1df82230; line: 7, column: 30] (through reference chain: data.pof.PortablePerson["dateOfBirth"])
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:148)
	at com.fasterxml.jackson.databind.DeserializationContext.wrongTokenException(DeserializationContext.java:943)
	at com.fasterxml.jackson.databind.jsontype.impl.AsArrayTypeDeserializer._locateTypeId(AsArrayTypeDeserializer.java:127)
	at com.fasterxml.jackson.databind.jsontype.impl.AsArrayTypeDeserializer._deserialize(AsArrayTypeDeserializer.java:93)
	at com.fasterxml.jackson.databind.jsontype.impl.AsArrayTypeDeserializer.deserializeTypedFromScalar(AsArrayTypeDeserializer.java:63)
	at com.fasterxml.jackson.databind.deser.std.StdScalarDeserializer.deserializeWithType(StdScalarDeserializer.java:26)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:518)
	at com.fasterxml.jackson.module.afterburner.deser.SettableObjectMethodProperty.deserializeAndSet(SettableObjectMethodProperty.java:52)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:256)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:159)
	at com.fasterxml.jackson.module.afterburner.deser.SuperSonicBeanDeserializer.deserialize(SuperSonicBeanDeserializer.java:121)
```
After looking into it a bit deeper, it turned out that the reason for this is this piece of code in `SettableBeanProperty.deserialize` method:

```
        if (_valueTypeDeserializer != null) {
            return _valueDeserializer.deserializeWithType(p, ctxt, _valueTypeDeserializer);
        }
```
Normally,` java.util.Date` deserialization should never hit this, but I have a mix-in defined that adds `@JsonTypeInfo` to `java.lang.Comparable`, and via inheritance to `java.util.Date` as well, so Jackson creates type deserializer and the code above ends up being called.

The problem is that none of the deserializers defined within `DateDeserializers` class override `deserializeWithType` method (unlike primitive deserializers), so the base `StdScalarDeserializer.deserializeWithType` method gets called instead.

I have the unit test that reproduces the issue and the fix in my fork and will submit pull request shortly.

**Comment by cowtowncoder** [comment]
@aseovic Thank you for reporting this, and especially for figuring out the root cause. I would love to get this resolved appropriately both for 2.7.1, and for 2.6 maintenance branch (for 2.6.6). Looking forward to PR.

One practical thing if and when you are ready for PR: if we have not yet gotten a filled-in Contributor License Agreement (CLA), we would need one. Default one is available at:

https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

and the easiest way is usually to print, fill+sign, scan and email to `info` at fasterxml dot com.
One copy is enough for all contributions; we just need it to ensure we can license all code as open source etc.

**Comment by aseovic** [comment]
I'm gonna close this, based on the discussion and conclusion in #1104.
